core/typedefs.h

```diff
@@ -37,60 +37,41 @@
  * Basic definitions and simple functions to be used everywhere.
  */
 
+// Include first in case the platform needs to pre-define/include some things.
 #include "platform_config.h"
 
+// Should be available everywhere.
+#include "core/error_list.h"
+#include "core/int_types.h"
+
+// Turn argument to string constant:
+// https://gcc.gnu.org/onlinedocs/cpp/Stringizing.html#Stringizing
 #ifndef _STR
 #define _STR(m_x) #m_x
 #define _MKSTR(m_x) _STR(m_x)
 #endif
 
-//should always inline no matter what
+// Should always inline no matter what.
 #ifndef _ALWAYS_INLINE_
-
-#if defined(__GNUC__) && (__GNUC__ >= 4)
-#define _ALWAYS_INLINE_ __attribute__((always_inline)) inline
-#elif defined(__llvm__)
+#if defined(__GNUC__)
 #define _ALWAYS_INLINE_ __attribute__((always_inline)) inline
 #elif defined(_MSC_VER)
 #define _ALWAYS_INLINE_ __forceinline
 #else
 #define _ALWAYS_INLINE_ inline
 #endif
-
 #endif
 
-//should always inline, except in some cases because it makes debugging harder
+// Should always inline, except in debug builds because it makes debugging harder.
 #ifndef _FORCE_INLINE_
-
 #ifdef DISABLE_FORCED_INLINE
 #define _FORCE_INLINE_ inline
 #else
 #define _FORCE_INLINE_ _ALWAYS_INLINE_
 #endif
-
 #endif
 
-//custom, gcc-safe offsetof, because gcc complains a lot.
-template <class T>
-T *_nullptr() {
-	T *t = NULL;
-	return t;
-}
-
-#define OFFSET_OF(st, m) \
-	((size_t)((char *)&(_nullptr<st>()->m) - (char *)0))
-/**
- * Some platforms (devices) don't define NULL
- */
-
-#ifndef NULL
-#define NULL 0
-#endif
```

Comment:
I haven't actually double checked if this is still necessary nowadays. But I think any compiler that supports C11 should define `NULL` in `typedefs.h` which we include above?

Reply:
I plan to make a PR migrating from NULL to nullptr so I guess it will be removed anyway. I'm waiting for the display-server branch to be merged so my PR doesn't cause a lot of conflicts.

Reply:
I was waiting similarly before doing this change, but if you want to work on it, I'll leave it to you for after the merge of the `display-server` branch.
Note that `clang-tidy` has a tool for that: https://clang.llvm.org/extra/clang-tidy/checks/modernize-use-nullptr.html